### DISCUSSION TITLE
Documentation Update Issues

### DISCUSSION THREAD

**Discussion by pavlis** [mspass-team/mspass#116] [Design & Development]
I'm in the middle of a rewrite of our existing user's manual.   I ran across an inconsistency.   I'm not sure if this is a gap in the implementation or my understanding.  

In the [database class documentation](https://wangyinz.github.io/mspass/python_api/mspasspy.db.html#module-mspasspy.db.database) is a description of a *read_distributed_data* method but there is no comparable *save_distributed_data*.   I think the idea you are thinking is this is a trivial application of map using the save_data method.  Is that right?  Nonetheless, I think for consistency we should define a *save_distributed_data* method.

**Comment by pavlis** [comment]
Here is another point that came up while writing documentation on the database.  Unclear to me from sphnyx documentation if read_data is dogmatic about loading source and site and/or channel data?  It absolutely most not be rigid on that point. There are many examples where one or the other is not required for a workflow. For example, the now common process of noise cross-correlation needs no source metadata, but probably does want site metadata.

**Comment by wangyinz** [comment]
Yeah, actually you revealed a hole in our original design. When we first came up with `read_distributed_data`, I was thinking it just a convenient way of turning a query result into a distributed data. Even without it, I thought the users can easily read or save data with the methods we defined in our `Database` class. Well, turns out that assumption was wrong because instances of the `Database` class are actually not serializable due to its internal lock. By design, the `Database` class shouldn't be duplicated, and multiple processes should instantiate their own client to the database anyway.

So, there are two ways to get around this: 1) define a number of wrapper functions similar to `read_distributed_data` that can do all database operations in parallel; 2) implement a customized serialization method such that the serialized `Database` will always create new client connection. The former is easier, but the latter seems more elegant. I definitely need to see how feasible that is.

For the `read_data`, I don't think it is dogmatic in current implementation. What it does should be reading the metadata specified in the schema.

**Comment by pavlis** [comment]
Roger that last comment about distributed data.

On the dogmatic issue it certainly isn't clear from the call sequence for read_data how it will handle normalized data.  When I look at the read_data function it isn't entirely clear to me how this works until I try doing something like stepping through it with a debugger.  I'm not at all clear how this does or does not handle normalized data in site, channel, or source.  It looks to me that at best the reader simply loads ids like site_id.  I'll know more tomorrow morning when I try running the current version on a prototype workflow.  

If the current stuff is hidden behind the Database handle or schema implementation I don't think that is the right design.  Seems to me the read_data function should be changed to this signature:
```
def read_data(object_id, object_type='TimeSeries', load_history=True, collection=None,
     normalize=['site','source']):
```
where the normalize parameter is a list of collection names the reader should try to query to fetch appropriate Metadata.  It is appropriate to have the name in the Schema object but we definitely need a cleaner way to control what is loaded.

**Comment by wangyinz** [comment]
I have implemented the custom pickle, and the users can save data like this:
https://github.com/wangyinz/mspass/blob/ea7a9b4b6fc036a3fba5d5c86585ab7e4a069e02/python/tests/db/test_db.py#L540-L542
or this:
https://github.com/wangyinz/mspass/blob/ea7a9b4b6fc036a3fba5d5c86585ab7e4a069e02/python/tests/db/test_db.py#L588-L590

Now, we don't need the save parallel data or any other similar functions, because it is more common and intuitive to directly call the map method of parallel data instead of applying function to them.

**Reply by pavlis** [reply]
That is slick and a good lesson.  You didn't say that, but it seems you could use those fragment to build a save_distributed_data.  Is that your plan?   If so, I'll put that in the revised documentation I'm beating on.

**Reply by wangyinz** [reply]
I am not sure if that's still necessary. The `save_distributed_data` will be as simple as:
```python
def save_distributed_data(pdata, db):
    if isinstance(pdata, pyspark.rdd.RDD):
        pdata.foreach(lambda d, db=db: db.save_data(d, 'gridfs'))
    if isinstance(pdate, dask.bag.core.Bag):
        pdata.map(db.save_data, 'gridfs').compute()
```
Note that there is a problem in such a function, which is the `foreach` call in spark and the `compute` call will start the actual computation immediately. This may not be the behavior the users wanted because if you only call the `map` or `reduce` methods, the computation will not start and the parallel engine is only building the task graph internally. They only start computing through the task graph when needed, which can be done through `persist` for Spark and `compute` for Dask. Note that this design of laziness is important to make such a parallel system run efficiently. By calling this save function, a user might unexpectedly launch a large parallel job. 

Alternatively, we can replace the `foreach` with `map` for Spark and remove the `compute` for Dask, so that the function call won't act on the data immediately, but that might be counterintuitive from a function's standpoint - the save function does not save the data on return. So, maybe we want to leave that option to the user, but it then complicates the interface because user will then have two different systems to determine when a task graph is launched: the Spark or Dask's native way and MsPASS' way. Since we are not going to hide the Spark or Dask's API from the users, we'd better not introduce another layer of complexity.

Note that the `read_distributed_data` don't have such a complexity, because the parallel data has to be initiated somewhere, and we are just providing an interface to initiate one from a database query. Once it is initiated, users can rely on the parallel engine's API to manipulate their data, and the functionalities of MsPASS operates at one level below the parallel engine - it does not directly interact with the data types of the parallel engine (such as RDD and Bag).

So, in the documentation, we can just show the two example lines above as the recommended way to call our save method within the parallel engine.

**Comment by wangyinz** [comment]
About the `read_data` interface, yes, we do fully rely on the metadata schema to specify what to be read or saved. I thought this is the design we started with. While it does appear counterintuitive at the beginning, it is actually more error proof. This is because a parameter like `normalize=['site','source']` is not enough to specify the accurate behavior of the reader (or writer). User will still want to specify what attributes within those collections are needed, and they will need to clarify what to do with the ones that are specified but missing or not specified by exist. We will either assume some default behavior, which will be even more confusing, or making the interface extremely verbose. 

I think that is exactly the reason why we designed the original `MetadataDefinitions` class. Later, we moved that class out of the C++ code, and re-implemented it as the `MetadataSchema` class. Also, because we were having the problem of mixing the concept of database and metadata, we now have them distinctively defined as `DatabaseSchema` and `MetadataSchema`. With the new schema class, users are supposed to specify exactly what is what in there, instead of passing in a number of arguments. 

While that is how we designed the database interface, I do agree it is confusing because we don't have the complete solution to let the user easily switch the read/write behavior within a workflow. The inventory we have right now includes the set methods to switch the schema for the `Database`:

https://github.com/wangyinz/mspass/blob/b5c9281b653d1f777eab8b00905d7a14bf997055/python/mspasspy/db/database.py#L111-L125

and the default related methods to specify which collections are the default ones to interact with: 

https://github.com/wangyinz/mspass/blob/b5c9281b653d1f777eab8b00905d7a14bf997055/python/mspasspy/db/schema.py#L253-L327
 
What is missing is that we don't have a method to change the default collection for TimeSeries or Seismograms defined in a `MetadataSchema`. I have mentioned this issue here: https://github.com/wangyinz/mspass/pull/98#discussion_r550845202, but @JiaoMaWHU and I decided to put that aside for that pull request. I guess it is now critical for us to think more about this, and resolve that issue. Please correct me if wrong @JiaoMaWHU, but I don't think it works if we simply specify a `collection` that is not defined in the schema at https://github.com/wangyinz/mspass/blob/b5c9281b653d1f777eab8b00905d7a14bf997055/python/mspasspy/db/database.py#L127

**Reply by JiaoMaWHU** [reply]
https://github.com/wangyinz/mspass/blob/b5c9281b653d1f777eab8b00905d7a14bf997055/python/mspasspy/db/database.py#L143-L150
https://github.com/wangyinz/mspass/blob/b5c9281b653d1f777eab8b00905d7a14bf997055/data/yaml/mspass.yaml#L263-L268
The wf_collection is defined here, it will be the collection of the '_id' attribute. So a user can change this attribute to reply on any collections he/she wants.

As for `collection`, I remember we discussed that we better leave this argument for convenience instead of removing it.

Does this answer the question? Any points I'm missing?

**Reply by wangyinz** [reply]
Yes, thanks for confirming that. Basically, now we need to add some feature to our schema class so it can adapt to different needs within a workflow.

**Reply by JiaoMaWHU** [reply]
Correct me if I am wrong, are we talking about maybe we want to skip attributes e.g. site attributes when reading an object? In that way, one can definitely add a `skip` attribute in the collections to skip them right? Just need to add some logic down here.
https://github.com/wangyinz/mspass/blob/b5c9281b653d1f777eab8b00905d7a14bf997055/python/mspasspy/db/database.py#L174-L182

**Reply by wangyinz** [reply]
You are correct. Actually this is a much more straightforward solution to the need that Gary described at the very beginning. I was really thinking it at a different level where users may want to start with one metadata schema and then change it slightly. In this case, having to create two different yaml files is not convenient, and we should provide user the ability to modify an existing schema.

**Comment by pavlis** [comment]
You are both deep in the weeds on this point on things the two of you worked out and I was only on the sidelines.  Can you summarize for me how I should describe how we handle normalized data in this database api?  That is the generic issue that prompted this discussion.

**Comment by wangyinz** [comment]
First of all, we need to implement a `exclude` argument for the `read_data` method as @JiaoMaWHU suggested. Then, we should think of what method should be added to the schema class to make it more flexible.

For the documentation, this is how we handle normalized data: 1) the normalized collections should be first defined under the `Database` section of the yaml file such that the corresponding entries to the `Metadata` section should exist; 2) the `Metadata` section defines what attributes will be read out of or write into the database; 3) the normalized data should follow the format of `col_attr` in the `Metadata` section, where the `col` is the collection name and the `attr` is the attribute name within that collection; 4) it is important that the attribute referenced like this should have the `col_attr` defined as its aliases because the read method relies on the alias mechanism to resolve `col_attr` into `attr`; 5) users can either use the `exclude` argument to override the behavior defined in the loaded yaml file or change the schema definition with `set_database_schema` and `set_metadata_schema`.

While writing above, I found that our save method does not handle normalized data at the moment. It will only insert the non-readonly keys defined in the `MetadataSchema` into the wf collection.

**Reply by JiaoMaWHU** [reply]
Well, I actually was proposing that maybe we can add a `skip` attribute in the YAML file, but as you said, this might add complexity.
